clippy_lints/src/manual_let_else.rs

```diff
@@ -0,0 +1,269 @@
+}
+
+impl ManualLetElse {
+    #[must_use]
+    pub fn new(msrv: Option<RustcVersion>) -> Self {
+        Self { msrv }
+    }
+}
+
+impl_lint_pass!(ManualLetElse => [MANUAL_LET_ELSE]);
+
+impl<'tcx> LateLintPass<'tcx> for ManualLetElse {
+    fn check_stmt(&mut self, cx: &LateContext<'_>, stmt: &'tcx Stmt<'tcx>) {
+        let if_let_or_match = if_chain! {
+            if meets_msrv(self.msrv, msrvs::LET_ELSE);
+            if !in_external_macro(cx.sess(), stmt.span);
+            if let StmtKind::Local(local) = stmt.kind;
+            if let Some(init) = local.init;
+            if !from_different_macros(init.span, stmt.span);
+            if let Some(if_let_or_match) = IfLetOrMatch::parse(cx, init);
+            then {
+                if_let_or_match
+            } else {
+                return;
+            }
+        };
+
+        match if_let_or_match {
+            IfLetOrMatch::IfLet(_let_expr, let_pat, if_then, if_else) => if_chain! {
+                if expr_is_simple_identity(let_pat, if_then);
+                if let Some(if_else) = if_else;
+                if expr_diverges(cx, if_else);
+                then {
+                    span_lint(
+                        cx,
+                        MANUAL_LET_ELSE,
+                        stmt.span,
+                        "this could be rewritten as `let else`",
+                    );
+                }
+            },
+            IfLetOrMatch::Match(_match_expr, arms, source) => {
+                if source != MatchSource::Normal {
+                    return;
+                }
+                // Any other number than two arms doesn't (neccessarily)
+                // have a trivial mapping to let else.
+                if arms.len() != 2 {
+                    return;
+                }
+                // We iterate over both arms, trying to find one that is an identity,
+                // one that diverges. Our check needs to work regardless of the order
+                // of both arms.
+                let mut found_identity_arm = false;
+                let mut found_diverging_arm = false;
+                for arm in arms {
+                    // Guards don't give us an easy mapping to let else
+                    if arm.guard.is_some() {
+                        return;
+                    }
+                    if expr_is_simple_identity(arm.pat, arm.body) {
+                        found_identity_arm = true;
+                    } else if expr_diverges(cx, arm.body) && pat_allowed_for_else(cx, arm.pat) {
+                        found_diverging_arm = true;
+                    }
+                }
+                if !(found_identity_arm && found_diverging_arm) {
+                    return;
+                }
+                span_lint(cx, MANUAL_LET_ELSE, stmt.span, "this could be rewritten as `let else`");
+            },
+        }
+    }
+
+    extract_msrv_attr!(LateContext);
+}
+
+fn expr_diverges(cx: &LateContext<'_>, expr: &'_ Expr<'_>) -> bool {
```

Comment:
I made this PR some time ago already, so my memory might not be perfect, but I think I remember trying something like that but the type checker did stuff that made it not work in practice, so I ended up with this heuristic which obviously isn't as nice. I think the problem was that the type checker takes the `!` and coerces it to the type of the "then" block, in the step ensuring that "then" and "else" blocks have the same type. So you have to look one level deeper, which this code is doing.

Reply:
As for the check in the `else` expression, that's different code I think from the check that ensures that all match arms have the same type (or if and else blocks).

Reply:
Oh it might be that https://github.com/rust-lang/rust/pull/100288 is doing what we need...

Reply:
I think you just want to do
```rust
tcx.is_ty_uninhabited_from(
    tcx.parent_module(expr.hir_id),
    expr_ty,
    cx.param_env,
)
```

Reply:
Uninhabitedness is not *precisely* what we need, as currently this fails to compile:

```Rust
enum Un {}

fn foo() -> Un {
    panic!()
}

fn main() {
    let Some(v) = Some(0) else { foo() };
}
```

```
error[E0308]: `else` clause of `let...else` does not diverge
 --> src/main.rs:8:32
  |
8 |     let Some(v) = Some(0) else { foo() };
  |                                ^^^^^^^^^ expected `!`, found enum `Un`
  |
  = note: expected type `!`
             found enum `Un`
  = help: try adding a diverging expression, such as `return` or `panic!(..)`
  = help: ...or use `match` instead of `let...else`
```

Even though `Un` is uninhabited.

Reply:
Maybe also check if the unadjusted type is an empty enum in addition to uninhabitedness :thinking:

Reply:
I've tried out @camsteffen 's suggestion and the tests fail. See the last two commits, or alternatively this patch:

<details>

```diff
From 58b5fe125547899aba2446b31d634a563ad0eec5 Mon Sep 17 00:00:00 2001
From: est31 <MTest31@outlook.com>
Date: Wed, 12 Oct 2022 02:30:18 +0200
Subject: [PATCH] Try the suggestion for the divergence check

---
 clippy_lints/src/manual_let_else.rs |  59 +-----------
 tests/ui/manual_let_else.stderr     | 135 ++++------------------------
 2 files changed, 20 insertions(+), 174 deletions(-)

diff --git a/clippy_lints/src/manual_let_else.rs b/clippy_lints/src/manual_let_else.rs
index c4c203b7db7..3eed2290742 100644
--- a/clippy_lints/src/manual_let_else.rs
+++ b/clippy_lints/src/manual_let_else.rs
@@ -2,7 +2,6 @@ use clippy_utils::diagnostics::span_lint_and_then;
 use clippy_utils::higher::IfLetOrMatch;
 use clippy_utils::source::snippet_opt;
 use clippy_utils::ty::is_type_diagnostic_item;
-use clippy_utils::visitors::for_each_expr;
 use clippy_utils::{meets_msrv, msrvs, peel_blocks};
 use if_chain::if_chain;
 use rustc_data_structures::fx::FxHashSet;
@@ -15,7 +14,6 @@ use rustc_session::{declare_tool_lint, impl_lint_pass};
 use rustc_span::symbol::sym;
 use rustc_span::Span;
 use serde::Deserialize;
-use std::ops::ControlFlow;
 
 declare_clippy_lint! {
     /// ### What it does
@@ -160,60 +158,9 @@ fn emit_manual_let_else(cx: &LateContext<'_>, span: Span, expr: &Expr<'_>, pat:
 }
 
 fn expr_diverges(cx: &LateContext<'_>, expr: &'_ Expr<'_>) -> bool {
-    fn is_never(cx: &LateContext<'_>, expr: &'_ Expr<'_>) -> bool {
-        if let Some(ty) = cx.typeck_results().expr_ty_opt(expr) {
-            return ty.is_never();
-        }
-        false
-    }
-    let mut does_diverge = false;
-    for_each_expr(expr, |ex| {
-        match ex.kind {
-            ExprKind::Continue(_) | ExprKind::Break(_, _) | ExprKind::Ret(_) => {
-                does_diverge = true;
-                ControlFlow::Break(())
-            },
-            ExprKind::Call(call, _) => {
-                if is_never(cx, ex) || is_never(cx, call) {
-                    does_diverge = true;
-                    return ControlFlow::Break(());
-                }
-                ControlFlow::Continue(())
-            },
-            ExprKind::MethodCall(..) => {
-                if is_never(cx, ex) {
-                    does_diverge = true;
-                    return ControlFlow::Break(());
-                }
-                ControlFlow::Continue(())
-            },
-            ExprKind::If(if_expr, if_then, if_else) => {
-                let else_diverges = if_else.map_or(false, |ex| expr_diverges(cx, ex));
-                let diverges = expr_diverges(cx, if_expr) || (else_diverges && expr_diverges(cx, if_then));
-                if diverges {
-                    does_diverge = true;
-                }
-                ControlFlow::Break(())
-            },
-            ExprKind::Match(match_expr, match_arms, _) => {
-                let diverges =
-                    expr_diverges(cx, match_expr) || match_arms.iter().all(|arm| expr_diverges(cx, arm.body));
-                if diverges {
-                    does_diverge = true;
-                }
-                ControlFlow::Break(())
-            },
-
-            // Don't continue into loops, as they are breakable,
-            // and we'd have to start checking labels.
-            ExprKind::Loop(..) => ControlFlow::Break(()),
-
-            // Default: descend
-            // TODO: support breakable blocks. For this we would need label support.
-            _ => ControlFlow::Continue(()),
-        }
-    });
-    does_diverge
+    let Some(ty) = cx.typeck_results().expr_ty_opt(expr) else { return false };
+    cx.tcx
+        .is_ty_uninhabited_from(cx.tcx.parent_module(expr.hir_id).into(), ty, cx.param_env)
 }
 
 fn pat_allowed_for_else(cx: &LateContext<'_>, pat: &'_ Pat<'_>, check_types: bool) -> bool {
diff --git a/tests/ui/manual_let_else.stderr b/tests/ui/manual_let_else.stderr
index e073276d20c..4d05cb223d6 100644
--- a/tests/ui/manual_let_else.stderr
+++ b/tests/ui/manual_let_else.stderr
@@ -1,11 +1,3 @@
-error: this could be rewritten as `let...else`
-  --> $DIR/manual_let_else.rs:17:5
-   |
-LL |     let v = if let Some(v_some) = g() { v_some } else { return };
-   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider writing: `let Some(v_some) = g() else { return };`
-   |
-   = note: `-D clippy::manual-let-else` implied by `-D warnings`
-
 error: this could be rewritten as `let...else`
   --> $DIR/manual_let_else.rs:18:5
    |
@@ -16,6 +8,7 @@ LL | |         return;
 LL | |     };
    | |______^
    |
+   = note: `-D clippy::manual-let-else` implied by `-D warnings`
 help: consider writing
    |
 LL ~     let Some(v_some) = g() else {
@@ -44,108 +37,6 @@ LL +         return;
 LL +     };
    |
 
-error: this could be rewritten as `let...else`
-  --> $DIR/manual_let_else.rs:37:9
-   |
-LL |         let v = if let Some(v_some) = g() { v_some } else { continue };
-   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider writing: `let Some(v_some) = g() else { continue };`
-
-error: this could be rewritten as `let...else`
-  --> $DIR/manual_let_else.rs:38:9
-   |
-LL |         let v = if let Some(v_some) = g() { v_some } else { break };
-   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider writing: `let Some(v_some) = g() else { break };`
-
-error: this could be rewritten as `let...else`
-  --> $DIR/manual_let_else.rs:42:5
-   |
-LL |     let v = if let Some(v_some) = g() { v_some } else { panic!() };
-   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider writing: `let Some(v_some) = g() else { panic!() };`
-
-error: this could be rewritten as `let...else`
-  --> $DIR/manual_let_else.rs:45:5
-   |
-LL | /     let v = if let Some(v_some) = g() {
-LL | |         v_some
-LL | |     } else {
-LL | |         std::process::abort()
-LL | |     };
-   | |______^
-   |
-help: consider writing
-   |
-LL ~     let Some(v_some) = g() else {
-LL +         std::process::abort()
-LL +     };
-   |
-
-error: this could be rewritten as `let...else`
-  --> $DIR/manual_let_else.rs:52:5
-   |
-LL | /     let v = if let Some(v_some) = g() {
-LL | |         v_some
-LL | |     } else {
-LL | |         if true { return } else { panic!() }
-LL | |     };
-   | |______^
-   |
-help: consider writing
-   |
-LL ~     let Some(v_some) = g() else {
-LL +         if true { return } else { panic!() }
-LL +     };
-   |
-
-error: this could be rewritten as `let...else`
-  --> $DIR/manual_let_else.rs:59:5
-   |
-LL | /     let v = if let Some(v_some) = g() {
-LL | |         v_some
-LL | |     } else if true {
-LL | |         return;
-LL | |     } else {
-LL | |         panic!("diverge");
-LL | |     };
-   | |______^
-   |
-help: consider writing
-   |
-LL ~     let Some(v_some) = g() else { if true {
-LL +         return;
-LL +     } else {
-LL +         panic!("diverge");
-LL +     } };
-   |
-
-error: this could be rewritten as `let...else`
-  --> $DIR/manual_let_else.rs:68:5
-   |
-LL | /     let v = if let Some(v_some) = g() {
-LL | |         v_some
-LL | |     } else {
-LL | |         match (g(), g()) {
-...  |
-LL | |         }
-LL | |     };
-   | |______^
-   |
-help: consider writing
-   |
-LL ~     let Some(v_some) = g() else {
-LL +         match (g(), g()) {
-LL +             (Some(_), None) => return,
-LL +             (None, Some(_)) => {
-LL +                 if true {
-LL +                     return;
-LL +                 } else {
-LL +                     panic!();
-LL +                 }
-LL +             },
-LL +             _ => return,
-LL +         }
-LL +     };
-   |
-
 error: this could be rewritten as `let...else`
   --> $DIR/manual_let_else.rs:85:5
    |
@@ -181,15 +72,23 @@ LL +     };
    |
 
 error: this could be rewritten as `let...else`
-  --> $DIR/manual_let_else.rs:101:13
+  --> $DIR/manual_let_else.rs:165:5
+   |
+LL | /     let v = if let Some(v_some) = None {
+LL | |         v_some
+LL | |     } else {
+LL | |         // Don't lint if the type is uninhabited but not !
+LL | |         un()
+LL | |     };
+   | |______^
    |
-LL |             let $n = if let Some(v) = $e { v } else { return };
-   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider writing: `let Some(v) = g() else { return };`
-...
-LL |     create_binding_if_some!(w, g());
-   |     ------------------------------- in this macro invocation
+help: consider writing
+   |
+LL ~     let Some(v_some) = None else {
+LL +         // Don't lint if the type is uninhabited but not !
+LL +         un()
+LL +     };
    |
-   = note: this error originates in the macro `create_binding_if_some` (in Nightly builds, run with -Z macro-backtrace for more info)
 
-error: aborting due to 13 previous errors
+error: aborting due to 5 previous errors
 

```

</details>

This lacks the empty enum check, so it is expected that the one false positive is added, but way more importantly, it now starts not firing in quite many instances, something the enum check would not help with. I think this matches my memory. Do you agree that we can go with my version of the code? Unless you have other suggestions.

Btw, running @camsteffen 's suggestion on the dogfood test *did* turn up one instance that the longer version of the function does not cover, so it was an useful excercise in the end:

<details>

```
diff --git a/clippy_lints/src/dereference.rs b/clippy_lints/src/dereference.rs
index 45ee2fce4..aba2a8a82 100644
--- a/clippy_lints/src/dereference.rs
+++ b/clippy_lints/src/dereference.rs
@@ -270,17 +270,15 @@ impl<'tcx> LateLintPass<'tcx> for Dereferencing<'tcx> {
             if let Some((state, data)) = self.state.take() {
                 report(cx, expr, state, data);
             }
             return;
         }
 
         let typeck = cx.typeck_results();
-        let (kind, sub_expr) = if let Some(x) = try_parse_ref_op(cx.tcx, typeck, expr) {
-            x
-        } else {
+        let Some((kind, sub_expr)) = try_parse_ref_op(cx.tcx, typeck, expr) else {
             // The whole chain of reference operations has been seen
             if let Some((state, data)) = self.state.take() {
                 report(cx, expr, state, data);
             }
             return;
         };
 
```

</details>

The issue here is with the descending behaviour I think, so I think it's fixable.